File: frontend/src/api/knowledge.ts
import { ApiError, apiRequest } from './client'

export interface KnowledgeSearchItem {
  source_id: string
  document_id: string
  article_id: string
  title: string
  category: string
  heading: string | null
  excerpt: string
  relevance_score: number
  source_path: string
}

export interface KnowledgeSearchResponse {
  query: string
  embedding_model: string
  items: KnowledgeSearchItem[]
}

function isSearchItem(value: unknown): value is KnowledgeSearchItem {
  if (typeof value !== 'object' || value === null) return false
  const item = value as Record<string, unknown>
  return (
    typeof item.source_id === 'string' &&
    typeof item.document_id === 'string' &&
    typeof item.article_id === 'string' &&
    typeof item.title === 'string' &&
    typeof item.category === 'string' &&
    (typeof item.heading === 'string' || item.heading === null) &&
    typeof item.excerpt === 'string' &&
    typeof item.relevance_score === 'number' &&
    typeof item.source_path === 'string'
  )
}

function validateSearchResponse(value: unknown): KnowledgeSearchResponse {
  if (typeof value !== 'object' || value === null) {
    throw new ApiError('ResolveAI returned an invalid knowledge response.')
  }
  const response = value as Record<string, unknown>
  if (
    typeof response.query !== 'string' ||
    typeof response.embedding_model !== 'string' ||
    !Array.isArray(response.items) ||
    !response.items.every(isSearchItem)
  ) {
    throw new ApiError('ResolveAI returned an invalid knowledge response.')
  }
  return response as unknown as KnowledgeSearchResponse
}

export async function searchKnowledge(query: string, signal?: AbortSignal) {
  const params = new URLSearchParams({ q: query, top_k: '5' })
  const response = await apiRequest<unknown>(
    `/knowledge/search?${params.toString()}`,
    { signal },
  )
  return validateSearchResponse(response)
}

File: backend/app/services/knowledge.py
from typing import Protocol
from uuid import UUID

from sqlalchemy.ext.asyncio import AsyncSession

from app.errors import NotFoundError
from app.models.knowledge import KnowledgeDocument
from app.rag.embeddings import EmbeddingProvider, LocalHashEmbedding
from app.rag.expansion import category_hints, expand_query
from app.rag.retrieval import RankedChunk, RetrievalCandidate, hybrid_rerank
from app.repositories.knowledge import KnowledgeRepository
from app.schemas.knowledge import (
    KnowledgeDocumentChunk,
    KnowledgeDocumentResponse,
    KnowledgeSearchItem,
    KnowledgeSearchResponse,
)


class RetrievalRepository(Protocol):
    async def vector_candidates(
        self,
        embedding: list[float],
        embedding_model: str,
        *,
        category: str | None,
        limit: int,
    ) -> list[RetrievalCandidate]: ...

    async def document(self, document_id: UUID) -> KnowledgeDocument | None: ...


class KnowledgeSearchService:
    def __init__(
        self,
        db: AsyncSession,
        embedder: EmbeddingProvider | None = None,
        repository: RetrievalRepository | None = None,
    ) -> None:
        self.embedder = embedder or LocalHashEmbedding()
        self.repository = repository or KnowledgeRepository(db)

    async def search(self, query: str, category: str | None, top_k: int) -> KnowledgeSearchResponse:
        expanded = expand_query(query)
        candidates = await self.repository.vector_candidates(
            self.embedder.embed(expanded),
            self.embedder.name,
            category=category,
            limit=min(200, max(40, top_k * 10)),
        )
        ranked = hybrid_rerank(
            expanded,
            candidates,
            top_k=top_k,
            preferred_categories=category_hints(query),
        )
        return KnowledgeSearchResponse(
            query=query,
            embedding_model=self.embedder.name,
            items=[self._search_item(item) for item in ranked],
        )

    async def get_document(self, document_id: UUID) -> KnowledgeDocumentResponse:
        document = await self.repository.document(document_id)
        if document is None:
            raise NotFoundError
        return self._document_response(document)

    @staticmethod
    def _search_item(item: RankedChunk) -> KnowledgeSearchItem:
        value = item.candidate
        return KnowledgeSearchItem(
            source_id=value.source_id,
            document_id=value.document_id,
            article_id=value.article_id,
            title=value.title,
            category=value.category,
            heading=value.heading,
            excerpt=value.content,
            relevance_score=round(item.relevance_score, 8),
            source_path=value.source_path,
        )

    @staticmethod
    def _document_response(document: KnowledgeDocument) -> KnowledgeDocumentResponse:
        return KnowledgeDocumentResponse(
            document_id=document.id,
            article_id=document.article_id,
            title=document.title,
            category=document.category,
            version=document.version,
            checksum=document.checksum,
            source_path=document.source_path,
            is_active=document.is_active,
            ingested_at=document.ingested_at,
            created_at=document.created_at,
            updated_at=document.updated_at,
            chunks=[
                KnowledgeDocumentChunk(
                    source_id=chunk.id,
                    chunk_index=chunk.chunk_index,
                    heading=chunk.heading,
                    content=chunk.content,
                    token_count=chunk.token_count,
                    embedding_model=chunk.embedding_model,
                )
                for chunk in sorted(document.chunks, key=lambda value: value.chunk_index)
            ],
        )
